tests/test_gene_association_file.py

```diff
@@ -0,0 +1,136 @@
+    assert output_path.exists(), "Output file not created."
+    df_out = pd.read_csv(output_path)
+    assert not df_out.empty, "Output DataFrame is empty."
+    for col in ASSOCIATION_COL_TYPES.keys():
+        assert col in df_out.columns, f"Column {col} missing in output."
+
+@pytest.mark.unit
+def test_column_types(temp_csv_files, evidence_mapping_df):
+    input_path, output_path = temp_csv_files
+    process_go_annotations(str(input_path), str(output_path), evidence_mapping_df)
+    df = pd.read_csv(output_path)
+    for column, expected_type in ASSOCIATION_COL_TYPES.items():
+        not_null = df[column].dropna()
+        if not not_null.empty:
+            mismatches = ~not_null.map(lambda x: isinstance(x, expected_type))
+            assert not mismatches.any(), f"Column {column} has incorrect types."
+
+@pytest.mark.unit
+def test_negated_logic(temp_csv_files, evidence_mapping_df):
+    input_path, output_path = temp_csv_files
+    process_go_annotations(str(input_path), str(output_path), evidence_mapping_df)
+    df = pd.read_csv(output_path)
+    assert df["negated"].any(), "Expected at least one negated is True row, it present NOT|enables"
+    assert (~df["negated"]).any(), "Expected at least one negated is False row, it present enables"
+
+# --- Helper function for evidence mapping test --- 
+@pytest.mark.unit
+def test_merge_and_fallback(annotation_df, evidence_mapping_df):
+    """Test the evidence mapping merge and fallback logic."""
+    merged = merge_evidence_mapping(annotation_df, evidence_mapping_df)
+    assert merged.loc[0, "evidence_type"] == "ECO:0000250", "Direct match failed."
+    assert merged.loc[1, "evidence_type"] == "ECO:0000307", "Fallback match failed."
+    assert "DB_Reference" not in merged.columns, "'DB_Reference' should have been dropped after merge."
+
+@pytest.mark.unit
+def test_no_fallback():
+    annotation_df = pd.DataFrame([{"Evidence_Code": "IEA", "publications": "PMID:999999"}])
+    evidence_df = pd.DataFrame([{"Evidence_Code": "EXP", "DB_Reference": "GO_REF:0000024", "evidence_type": "ECO:0000269"}])
+    merged = merge_evidence_mapping(annotation_df, evidence_df)
+    assert pd.isna(merged.loc[0, "evidence_type"]), ("Expected evidence_type to be NaN when no match and no fallback are available.")
+
+@pytest.mark.unit
+def test_fallback_after_explode():
+    annotation_df = pd.DataFrame([{"Evidence_Code": "ND", "publications": ["PMID:111111", "PMID:222222"]}])
+    evidence_df = pd.DataFrame([{"Evidence_Code": "ND", "DB_Reference": "DEFAULT", "evidence_type": "ECO:0000307"}])
+    merged = merge_evidence_mapping(annotation_df, evidence_df)
+    assert len(merged) == 2, "Expected two rows after exploding publication list"
+    assert all(merged["evidence_type"] == "ECO:0000307"), "Fallback evidence_type not correctly applied after explode"
+
+@pytest.mark.unit
+def test_normalize_dates():
+    df = pd.DataFrame({"annotation_date": ["20200101", "20221301", "abcd1234", "20181231", None]})
```

Comment:
have you used `pytest.mark.parametrize` before? It provides a nice way to do these kinds of repeated tests.

No need to change this, just mentioning it for future reference. :)